pwndbg/commands/cymbol.py

```diff
@@ -0,0 +1,269 @@
+    except subprocess.CalledProcessError as exepction:
+        print(message.error(exepction))
+        print(message.error("Parsing failed, try to fix any syntax errors first."))
+        return None
+    except Exception as exepction:
+        print(message.error(exepction))
+        print(message.error("An error occured while generating the debug symbols."))
+        return None
+
+    return pwndbg_debug_symbols_output_file
+
+
+@PromptForOverwrite
+def add_custom_structure(custom_structure_name):
+    print(
+        message.notice("Enter your custom structure in a C header style, press Ctrl+D to save:\n")
+    )
+
+    custom_structures_source = sys.stdin.read().strip()
+    if custom_structures_source == "":
+        print(message.notice("An empty structure is entered, skipping ..."))
+        return
+
+    pwndbg_cachedir = pwndbg.lib.tempfile.cachedir("custom-symbols")
+    pwndbg_custom_structure_path = os.path.join(pwndbg_cachedir, custom_structure_name) + ".c"
+
+    with open(pwndbg_custom_structure_path, "w") as f:
+        f.write(custom_structures_source)
+
+    pwndbg_debug_symbols_output_file = generate_debug_symbols(pwndbg_custom_structure_path)
+    if not pwndbg_debug_symbols_output_file:
+        return
+
+    gdb.execute(f"add-symbol-file {pwndbg_debug_symbols_output_file}", to_string=True)
+    print(message.success("Symbols are added!"))
+
+
+@OnlyWhenStructureExists
+def edit_custom_structure(custom_structure_name, custom_structure_path):
+
+    # Lookup an editor to use for editing the custom structure.
+    editor_preference = os.getenv("EDITOR")
+    if not editor_preference:
+        editor_preference = os.getenv("VISUAL")
+    if not editor_preference:
+        editor_preference = "vi"
+
+    if cymbol_editor != "":
+        editor_preference = cymbol_editor
+
+    try:
+        subprocess.run(
+            [editor_preference, custom_structure_path],
+            stdout=subprocess.PIPE,
+            stderr=subprocess.PIPE,
+            check=True,
+        )
+    except Exception as exepction:
+        print(message.error("An error occured during opening the source file."))
+        print(message.error(f"Path to the custom structure: {custom_structure_path}"))
+        print(message.error("Please try to manually edit the structure."))
+        print(
+            message.error(
+                '\nTry to set a path to an editor with:\n\tset "cymbol-editor" /usr/bin/nano'
+            )
+        )
+        return
+
+    input(message.notice("Press enter to continue."))
+
+    load_custom_structure(custom_structure_name)
+
+
+@OnlyWhenStructureExists
+def remove_custom_structure(custom_structure_name, custom_structure_path):
+    os.remove(custom_structure_path)
+    print(message.success("Symbols are removed!"))
+    print(
+        message.notice(
+            "If the symbols are already loaded, please restart pwndbg to unload them from memory."
```

Comment:
Can we try using https://stackoverflow.com/a/19681769/1508881 for that?

Reply:
Yeah I'm planning that when I will properly store the debug symbols in a `.local` directory. For now I think won't work because the file paths to the loaded generated symbols are randomly generated temporary files in `/tmp`.

Reply:
We can save/cache the paths in some dict like: `symbols[symbol_name] = symbol_file_path` and then we will have it here :)

Reply:
Good suggestion, also we will need `remove-symbol-file` after all for unloading the symbols each time we edit an already loaded symbol. It seems that multiple edits don't work currently for that reason.